Support for Messenger HandleTrait return types

bnowak opened the pull request:
The goal is to cover https://github.com/phpstan/phpstan-symfony/issues/207.

That's my very first PR in phpstan-symfony plugin, so any feedback/tips/advices are very welcome :)

Commit:
base version of handling query result

Commit:
coding standards fix

Commit:
phpstan fix

Commit:
phpstan fix 2

ondrejmirtes reviewed (changes requested):
Instead of DynamicMethodReturnTypeExtension you can use https://apiref.phpstan.org/1.11.x/PHPStan.Type.ExpressionTypeResolverExtension.html which is like that but lets you override any expression type.

Commit:
add /.idea path to .gitignore

Commit:
use ExpressionTypeResolverExtension instead of DynamicMethodReturnTypeExtension

Commit:
adjust MessageMap to contain only supported cases

Commit:
covered more test-cases

Commit:
get rid of todo comments in \PHPStan\Symfony\MessageMapFactory::guessHandledMessages

Commit:
addressed rest of todo comments

bnowak commented:
@ondrejmirtes Thanks for the tip about `ExpressionTypeResolverExtension`! It resolved my blocker issue about not supported traits in `DynamicMethodReturnTypeExtension`.

It took me some time, however finally I've got ready to review version of this PR :) Please take a look in you free time :)

Thank you very much!

bnowak commented:
As static analyzing should works now fine with the result of `HandleTrait::handle`, but only in classes which have statically typed query passed - like in this SF-adjusted example:
```php
// src/Action/ListItems.php
namespace App\Action;

use App\Message\ListItemsQuery;
use App\MessageHandler\ListItemsQueryResult;
use Symfony\Component\Messenger\HandleTrait;
use Symfony\Component\Messenger\MessageBusInterface;

class ListItems
{
    use HandleTrait;

    public function __construct(
        private MessageBusInterface $messageBus,
    ) {
    }

    public function __invoke(): void
    {
        $result = $this->handle(new ListItemsQuery(/* ... */));
        // $result should be automatically recognized as `ListItemsQueryResult` now by phpstan via SF configuration

        // Do something with the result
        // ...
    }
}
```

However, my target goal was to cover QueryBus classes with dynamic return types based on dynamic query, something like:
```php
// src/MessageBus/QueryBus.php
namespace App\MessageBus;

use Symfony\Component\Messenger\Envelope;
use Symfony\Component\Messenger\HandleTrait;
use Symfony\Component\Messenger\MessageBusInterface;

class QueryBus
{
    use HandleTrait;

    public function __construct(
        private MessageBusInterface $messageBus,
    ) {
    }

    // wherever that method is called, phpstan should recognize return type based on passed query
    public function query($query)
    {
        return $this->handle($query);
    }
}
```

Maybe it's trivial and doable with already existing php-doc annotation or more complex to write some next extension for that, but I don't know how to connect the dots and reuse covered typing of `HandleTrait::handle` which is already in place. I'm happy to address that in separate PR (if will be needed to achieve that) :)
Any thoughts/ideas are more than welcome :)
/cc @ondrejmirtes

KrzysztofCzerwinski commented on src/Symfony/MessageMap.php:
```diff
@@ -0,0 +1,26 @@
+final class MessageMap
+{
+
+	/** @var array<string, Type> */
```
suggestion: `class-string`

bnowak replied:
thanks, adjusted

KrzysztofCzerwinski commented on src/Symfony/MessageMap.php:
```diff
@@ -0,0 +1,26 @@
+	private $messageMap;
+
+	/**
+	 * @param array<string, Type> $messageMap
```
suggestion: `class-string`

bnowak replied:
thanks, adjusted

KrzysztofCzerwinski commented on src/Symfony/MessageMapFactory.php:
```diff
@@ -0,0 +1,120 @@
+		foreach ($this->serviceMap->getServices() as $service) {
+			$serviceClass = $service->getClass();
+
+			if (is_null($serviceClass)) {
```
suggestion: native null comaprison- a tiny bit faster

bnowak replied:
thanks, adjusted

KrzysztofCzerwinski commented on src/Symfony/MessageMapFactory.php:
```diff
@@ -0,0 +1,120 @@
+		}
+
+		$variants = $methodReflection->getVariants();
+		if (count($variants) !== 1) {
+			return;
+		}
+
+		$parameters = $variants[0]->getParameters();
+
+		if (count($parameters) !== 1) {
+			return;
+		}
+
+		$classNames = $parameters[0]->getType()->getObjectClassNames();
+
+		if (count($classNames) !== 1) {
+			return;
+		}
```
suggestion: `1 !== count(...)`- not exactly a case of yoda, but seems a bit more in yoda nature where the constant value is on the left hand side

bnowak replied:
adjusted, however btw. real yoda cases are disabled in this repo in coding standards :wink:

KrzysztofCzerwinski commented on src/Type/Symfony/MessengerHandleTraitReturnTypeExtension.php:
```diff
@@ -0,0 +1,85 @@
+
+	private function getMessageMap(): MessageMap
+	{
+		if (is_null($this->messageMap)) {
+			$this->messageMap = $this->messageMapFactory->create();
+		}
+
+		return $this->messageMap;
```
suggestion: either native comparison or:
```php
$this->messageMap = $this->messageMap ?? $this->messageMapFactory->create();

return $this->messageMap;
```

bnowak replied:
thanks, adjusted

Commit:
code review 1

Commit:
code review 2 removed mixed type which doesn't exist in <PHP8

bnowak commented:
@ondrejmirtes any chances to review that, please? :pray:

bnowak commented:
@ondrejmirtes I can see that version 2.0.0 is released now. Is 1.x abandoned from now? Should I rebase that PR to 2.x branch? Any chances to review and/or merge it please? :slightly_smiling_face:

ondrejmirtes commented on .gitignore:
```diff
@@ -1,5 +1,6 @@
 /tests/tmp
 /build-cs
 /vendor
+/.idea
```
Remove this please. This belongs to a global .gitignore.

ondrejmirtes commented on src/Symfony/MessageMapFactory.php:
```diff
@@ -0,0 +1,143 @@
+
+				/** @var array{handles?: class-string, method?: string} $tagAttributes */
+				$tagAttributes = $tag->getAttributes();
+				$reflectionClass = $this->reflectionProvider->getClass($serviceClass);
```
This is unsafe. Ask for hasClass first.

ondrejmirtes commented on src/Symfony/MessageMapFactory.php:
```diff
@@ -0,0 +1,143 @@
+				}
+
+				foreach ($handles as $messageClassName => $options) {
+					$methodReflection = $reflectionClass->getNativeMethod($options['method'] ?? self::DEFAULT_HANDLER_METHOD);
```
This is unsafe. Ask for hasNativeMethod first.

ondrejmirtes commented on src/Symfony/MessageMapFactory.php:
```diff
@@ -0,0 +1,143 @@
+	private function guessHandledMessages(ClassReflection $reflectionClass): iterable
+	{
+		if ($reflectionClass->implementsInterface(MessageSubscriberInterface::class)) {
+			foreach ($reflectionClass->getName()::getHandledMessages() as $index => $value) {
```
Save the class name into a variable first, then call the method.

ondrejmirtes commented on src/Symfony/MessageMapFactory.php:
```diff
@@ -0,0 +1,143 @@
+		return new MessageMap($messageMap);
+	}
+
+	/** @return array<class-string, array<string, string>> */
```
This does not return an array. It's a generator.

bnowak replied:
right, changed to iterable :)

ondrejmirtes commented on src/Symfony/MessageMapFactory.php:
```diff
@@ -0,0 +1,143 @@
+
+		try {
+			$methodReflection = $reflectionClass->getNativeMethod(self::DEFAULT_HANDLER_METHOD);
+		} catch (MissingMethodFromReflectionException $e) {
```
Ask for hasNativeMethod first, don't catch the exception.

ondrejmirtes commented on src/Symfony/MessageMapFactory.php:
```diff
@@ -0,0 +1,143 @@
+			return;
+		}
+
+		/** @var class-string[] $classNames */
```
Don't add this `@var`, this assumption might be wrong.

ondrejmirtes commented on src/Symfony/MessageMapFactory.php:
```diff
@@ -0,0 +1,143 @@
+			return false;
+		}
+
+		throw new ShouldNotHappenException();
```
I'd say this can happen very easily. Wrong getHandledMessages implementation should not crash PHPStan.

bnowak replied:
Right, so now I'm catching that case and skipping when wrongly implemented method.

ondrejmirtes commented on src/Symfony/MessageMapFactory.php:
```diff
@@ -0,0 +1,143 @@
+				}
+
+				/** @var array{handles?: class-string, method?: string} $tagAttributes */
+				$tagAttributes = $tag->getAttributes();
```
Can we add a `@return` stub for this method instead of `@var`?

bnowak replied:
This shape is dynamic and rely on what tag we're using from SF config. In this case I'm ensuring above that we're handling only `messenger.message_handler`, so we know what shape it should have.

As far I know stubs are static only, so unfortunately we cannot use them here.

Do you have other idea or it could stay as it is?

ondrejmirtes commented on src/Symfony/MessageMapFactory.php:
```diff
@@ -0,0 +1,143 @@
+	 * @phpstan-assert-if-true class-string $index
+	 * @phpstan-assert-if-true array<string, mixed> $value
+	 * @phpstan-assert-if-false int $index
+	 * @phpstan-assert-if-false class-string $value
```
All of these asserts need `=`: https://phpstan.org/writing-php-code/narrowing-types#equality-assertions

ondrejmirtes commented on src/Type/Symfony/MessengerHandleTraitReturnTypeExtension.php:
```diff
@@ -0,0 +1,86 @@
+		if ($this->isSupported($expr, $scope)) {
+			$arg = $expr->getArgs()[0]->value;
+			/** @var class-string[] $argClassNames */
+			$argClassNames = $scope->getType($arg)->getObjectClassNames();
```
Don't add this `@var`.

ondrejmirtes commented on src/Type/Symfony/MessengerHandleTraitReturnTypeExtension.php:
```diff
@@ -0,0 +1,86 @@
+	}
+
+	/**
+	 * @phpstan-assert-if-true MethodCall $expr
```
This needs `=`.

ondrejmirtes commented on src/Type/Symfony/MessengerHandleTraitReturnTypeExtension.php:
```diff
@@ -0,0 +1,86 @@
+
+			return $declaringClassReflection->getName() === self::TRAIT_NAME;
+		} catch (ReflectionException $e) {
+			return false;
```
Instead of catching exception, check if the method exists first.

ondrejmirtes reviewed (changes requested):
No need to rebase this, we could release this in 1.x.

Commit:
code review fixes

bnowak commented:
@ondrejmirtes Thanks for review. I addressed all your comments, please take a look again when you have time :)

Also, could you take a look on https://github.com/phpstan/phpstan-symfony/pull/404#issuecomment-2419761831 and answer that if you have an idea how to solve it please? That's my initial goal to resolve by this PR :)

ondrejmirtes commented on src/Symfony/MessageMapFactory.php:
```diff
@@ -0,0 +1,158 @@
+					} else {
+						yield $value => ['method' => self::DEFAULT_HANDLER_METHOD];
+					}
+				} catch (ShouldNotHappenException $e) {
```
Never catch ShouldNotHappenException.

ondrejmirtes commented on src/Type/Symfony/MessengerHandleTraitReturnTypeExtension.php:
```diff
@@ -0,0 +1,86 @@
+	public function getType(Expr $expr, Scope $scope): ?Type
+	{
+		if ($this->isSupported($expr, $scope)) {
+			$arg = $expr->getArgs()[0]->value;
```
`$expr->getArgs()[0]` might not exist. Check the existence first.

Commit:
code review fixes 2

bnowak commented:
@ondrejmirtes both comments fixed. Could you re-check? Thanks, in advance :)

ondrejmirtes commented:
Thank you. Please be available if any issues arise.

bnowak commented:
@ondrejmirtes Thank you for merging and your help with reviewing that :)

Could you take a look on https://github.com/phpstan/phpstan-symfony/pull/404#issuecomment-2419761831 comment please 🙏 I'm not sure whether I should create separate issue for that? Maybe it's able to achive with already existing features/annotations? If not, do you have an idea how to implement such feature?

ondrejmirtes commented:
@bnowak Feel free to send a failing test so I can understand it better.

ondrejmirtes commented:
Hi, since MessageSubscriberInterface has been deprecated/removed in favour of `#[AsMessageHandler]`, shouldn't the code in MessageMapFactory be adjusted to also support the attribute? Thanks.

bnowak commented:
@ondrejmirtes I'll take a look on that tomorrow. Thanks for info :wink:

ondrejmirtes commented:
Maybe it's the reason why some of the tests are failing on Symfony 7.3, maybe not: https://github.com/phpstan/phpstan-symfony/actions/runs/16193692378/job/45715050598?pr=447

bnowak commented:
Yes it is. I started working on tests separation and skipping the case when `MessageSubscriberInterface` is missing (SF <7.0). Tomorrow, I'll push PR for that.

Regarding adding support for `#[AsMessageHandler]` - I'll handle that in separate PR, no problem. However, I think firstly before that, it would be nice to have support for SF 7.3 in tests (work which you started in https://github.com/phpstan/phpstan-symfony/pull/447/files). Then, I'd add that support.

@ondrejmirtes What do you think? Does such order make sense?

ondrejmirtes commented:
I think AsMessageHandler already existed before Symfony 7, didn't it?

It'd make it easy to implement it here, and when my symfony-7-3 is rebased, less tests should fail.

bnowak commented:
Ahh, you're right. I don't know why I assumed it was introduced in 7.0 😅

Tomorrow will try to handle both topics 😉

bnowak commented:
@ondrejmirtes please review https://github.com/phpstan/phpstan-symfony/pull/448.

Regarding `AsMessageHandler` support - it's just tag for auto configuring it during kernel boot, so its result is generated container xml file definition which this plugin read from (the same as messenger handle trait feature).
Given that, it should be supported by default and additional tests are not needed IMO.
